### PR TITLE
add analytics to modals and selects

### LINKED ISSUE #1522: Add tracking calls for analytics
add ga `analytics.track(` calls for #1521

### PR DESCRIPTION
**Pull request by na3d**
Closes #1522

### CODE REVIEW

File: src/actions/trackEvent.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export function TrackEvent (cate, action, label, value = null) {
```
Comment (mparke) [inline src/actions/trackEvent.js]: Can we rename this from `TrackEvent` to something generic? `CustomEvent` maybe?

The purpose of dispatching an event is to decouple analytics from app logic/components. If it's called trackevent inside of generic components, there's no decoupling

Also, any place in `/components` where a propType has been added to be used in `trackEvent` also breaks this convention, that prop serves no purpose in the functionality of the component. In the example of modals, we may be better off calling this method from outside the component, in `onOk` `onCancel` called by whichever view is using the modal.

happy to assist with examples, let me know

File: src/index.js [hunk 2]
```diff
@@ -115,6 +115,17 @@ window.handleError = function (e) {
   return null;
 };
 
+document.addEventListener('track', function(e) {
```
Comment (mparke) [inline src/index.js]: looking good!

minor thing, can we get full var names here? `cate` means nothing to me hah

File: src/actions/trackEvent.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+  document.dispatchEvent(new CustomEvent('track', {
```
Comment (mparke) [inline src/actions/trackEvent.js]: wrap `dispatchEvent` in a check here for compatibility,

`if (document.dispatchEvent) {`

might be some compat issues with `CustomEvent` seen here: https://caniuse.com/#search=customevent
I'll find an appropriate check

File: components/forms/Select.js [hunk 4]
```diff
@@ -10,7 +11,11 @@ export default function Select(props) {
         name={props.name}
         id={props.id}
         disabled={props.disabled}
-        onChange={props.onChange}
+        onChange={(e) => {
+          const value = String(e.target.value);
```
Comment (mparke) [inline components/forms/Select.js]: why is `String()` needed here?

Comment (na3d) [inline components/forms/Select.js]: That was the solution to the error I was running into.  It seems odd, but who am I to question the tests? :p

File: src/actions/trackEvent.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+export function TrackEvent(cate, action, label, value = null) {
+  document.dispatchEvent(new CustomEvent('track', {
+    detail: {
+      cate,
```
Comment (abemassry) [inline src/actions/trackEvent.js]: what does cate refer to?

File: src/actions/trackEvent.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+export function TrackEvent(cate, action, label, value = null) {
```
Comment (abemassry) [inline src/actions/trackEvent.js]: what does cate refer to?

File: src/actions/trackEvent.js (hunk 5 above)
Comment (na3d) [inline src/actions/trackEvent.js]: Category

File: components/forms/Select.js (hunk 4 above)
Comment (mparke) [inline components/forms/Select.js]: you shouldn't need it, I can pull your branch and check

File: src/actions/trackEvent.js (hunk 6 above)
Comment (mparke) [inline src/actions/trackEvent.js]: I guess my previous comment got lost in an update, can we rename this from `TrackEvent` to `CustomEvent` ?

Comment (mparke) [inline src/actions/trackEvent.js]: Also I had comments about compatability, if this works in Chrome and FF for starters that's okay, I'll create follow ups to check

File: src/index.js [hunk 7]
```diff
@@ -125,6 +125,16 @@ window.handleError = function (e) {
   return null;
 };
 
+document.addEventListener('track', function (e) {
```
Comment (mparke) [inline src/index.js]: like I said before the purpose of using the custom event was to decouple analytics from implementation.

if the method is called `TrackEvent` and the event is called `track` there's no real decoupling, except for maybe  writing our own interface around `window.ga`

for example, I would expect `onCancel={function() { CustomEvent('click:cancel', something, something); }` < where nothing here is referencing analytics, just that something happened.

**Comment by mparke** [comment]
changes look fine, I would just like to rename a couple things so that analytics is not referenced in generic components.

File: src/actions/trackEvent.js (hunk 6 above)
Comment (na3d) [inline src/actions/trackEvent.js]: Wouldn't naming it CustomEvent override the CustomEvent on the next line?

I mainly didn't to future proof the possibilities of creating more CustomEvents in the future.

File: components/forms/Select.js (hunk 4 above)
Comment (na3d) [inline components/forms/Select.js]: Code wise, running as a program, it doesn't need it.  Test wise, it is requiring it for the tests to pass.

File: src/actions/trackEvent.js (hunk 6 above)
Comment (mparke) [inline src/actions/trackEvent.js]: oops good catch

**Comment by mparke** [comment]
Follow up: https://github.com/linode/manager/issues/2151

**Pull request merged**